### ISSUE TITLE
Enhance APICompat to discover [Serializable]

### ISSUE THREAD

**Issue by danmosemsft** [dotnet/corefx#12725]
Presumably some of these are missing, since BinarySerialization had been removed. It's hard to say how many because it's not represented in the ref assemblies. This may not be much (or even any) since there are still 149 instances of it in the CoreCLR implementations and we inherit most of those implementations. We do need it back since it likely changes serialization behavior.

To find them if any, we need to update the apicompat tool then run it against desktop impl on the LHS and CoreFX impl on the RHS. Any types that are present in CoreFX, but are missing the attribute, should have it added.

@weshaggard sound about right? perhaps there's no much to do here, since it's all over CoreCLR.

**Comment by AlexGhiondea** [comment]
@danmosemsft should this be moved to the buildtools repo?

**Comment by danmosemsft** [comment]
If you like -- figured we could track all NS2.0 work together. Up to you

**Comment by stephentoub** [comment]
> This may not be much (or even any) since there are still 149 instances of it in the CoreCLR implementations and we inherit most of those implementations. We do need it back since it likely changes serialization behavior.

When I'd checked a few months ago, we were missing ~200 across corefx.

**Comment by danmosemsft** [comment]
How did you figure that out? Still have the utility around?

**Comment by stephentoub** [comment]
> How did you figure that out?

I'd written a little tool to find all of the types in corefx that weren't serializable, and then used reflection over desktop to find all types that were.  I don't think I have the code around anymore, but it shouldn't be hard to recreate.  Though it may not be necessary; this is an old and never-intended-to-be-merged/never-compiled branch, but it provides a sort-of diff from the time of the serialization attributes/implementations we were missing:
https://github.com/stephentoub/corefx/commit/898da041950f5e25c4e2e3d788b6b6bd5fa78bc3

**Comment by stephentoub** [comment]
(I'm working on bringing over some of these, namely where the projects have already been upgraded to support 1.7.)

**Comment by stephentoub** [comment]
I wrote a new tool this morning that uses Roslyn to look over all of our source files finding non-[Serializable] types, uses reflection to look over all of the desktop assemblies looking for [Serializable] types, and then intersects them to get a list of the types that aren't [Serializable] in corefx but are in desktop.  Below are the results.  There are a few false positives here, e.g. because some assembly currently stubs out a type of the same full name as the one in question, but this should be fairly representative:
- [ ] Microsoft.CSharp.RuntimeBinder.RuntimeBinderException
- [ ] Microsoft.CSharp.RuntimeBinder.RuntimeBinderInternalCompilerException
- [ ] System.Threading.Tasks.Dataflow.DataflowMessageHeader
- [ ] System.DBNull
- [ ] System.OperationCanceledException
- [ ] System.PlatformID
- [ ] System.IO.BinaryWriter
- [ ] System.IO.DriveType
- [ ] System.IO.DriveInfo
- [ ] System.IO.DriveNotFoundException
- [ ] System.IO.EndOfStreamException
- [ ] System.IO.FileAccess
- [ ] System.IO.MemoryStream
- [ ] System.IO.StreamReader
- [ ] System.IO.StreamWriter
- [ ] System.IO.StringReader
- [ ] System.IO.StringWriter
- [ ] System.IO.TextReader
- [ ] System.IO.TextWriter
- [ ] System.IO.IsolatedStorage.IsolatedStorageScope
- [ ] System.Security.SecurityElement
- [ ] System.Security.XmlSyntaxException
- [ ] System.Security.CodeAccessPermission
- [ ] System.Security.HostSecurityManagerOptions
- [ ] System.Security.HostSecurityManager
- [ ] System.Security.NamedPermissionSet
- [ ] System.Security.PermissionSet
- [ ] System.Security.HostProtectionException
- [ ] System.Security.PolicyLevelType
- [ ] System.Security.SecurityZone
- [ ] System.Security.AccessControl.PrivilegeNotHeldException
- [ ] System.Security.Cryptography.CipherMode
- [ ] System.Security.Cryptography.PaddingMode
- [ ] System.Security.Cryptography.CryptographicException
- [ ] System.Security.Cryptography.CryptographicUnexpectedOperationException
- [ ] System.Security.Cryptography.FromBase64TransformMode
- [ ] System.Security.Cryptography.CspProviderFlags
- [ ] System.Security.Cryptography.CryptoStreamMode
- [ ] System.Security.Cryptography.KeyNumber
- [ ] System.Security.Cryptography.X509Certificates.X509KeyStorageFlags
- [ ] System.Security.Cryptography.X509Certificates.X509Certificate
- [ ] System.Security.Permissions.EnvironmentPermissionAccess
- [ ] System.Security.Permissions.EnvironmentPermission
- [ ] System.Security.Permissions.FileDialogPermissionAccess
- [ ] System.Security.Permissions.FileDialogPermission
- [ ] System.Security.Permissions.FileIOPermissionAccess
- [ ] System.Security.Permissions.FileIOPermission
- [ ] System.Security.Permissions.HostProtectionResource
- [ ] System.Security.Permissions.HostProtectionAttribute
- [ ] System.Security.Permissions.PermissionState
- [ ] System.Security.Permissions.SecurityAction
- [ ] System.Security.Permissions.SecurityAttribute
- [ ] System.Security.Permissions.CodeAccessSecurityAttribute
- [ ] System.Security.Permissions.EnvironmentPermissionAttribute
- [ ] System.Security.Permissions.FileDialogPermissionAttribute
- [ ] System.Security.Permissions.PrincipalPermissionAttribute
- [ ] System.Security.Permissions.ReflectionPermissionAttribute
- [ ] System.Security.Permissions.RegistryPermissionAttribute
- [ ] System.Security.Permissions.SecurityPermissionAttribute
- [ ] System.Security.Permissions.UIPermissionAttribute
- [ ] System.Security.Permissions.ZoneIdentityPermissionAttribute
- [ ] System.Security.Permissions.StrongNameIdentityPermissionAttribute
- [ ] System.Security.Permissions.SiteIdentityPermissionAttribute
- [ ] System.Security.Permissions.UrlIdentityPermissionAttribute
- [ ] System.Security.Permissions.PublisherIdentityPermissionAttribute
- [ ] System.Security.Permissions.PermissionSetAttribute
- [ ] System.Security.Permissions.ReflectionPermissionFlag
- [ ] System.Security.Permissions.ReflectionPermission
- [ ] System.Security.Permissions.PrincipalPermission
- [ ] System.Security.Permissions.SecurityPermissionFlag
- [ ] System.Security.Permissions.SecurityPermission
- [ ] System.Security.Permissions.SiteIdentityPermission
- [ ] System.Security.Permissions.StrongNameIdentityPermission
- [ ] System.Security.Permissions.StrongNamePublicKeyBlob
- [ ] System.Security.Permissions.UIPermissionWindow
- [ ] System.Security.Permissions.UIPermissionClipboard
- [ ] System.Security.Permissions.UIPermission
- [ ] System.Security.Permissions.UrlIdentityPermission
- [ ] System.Security.Permissions.ZoneIdentityPermission
- [ ] System.Security.Permissions.GacIdentityPermissionAttribute
- [ ] System.Security.Permissions.GacIdentityPermission
- [ ] System.Security.Permissions.PublisherIdentityPermission
- [ ] System.Security.Permissions.RegistryPermissionAccess
- [ ] System.Security.Permissions.RegistryPermission
- [ ] System.Security.Claims.Claim
- [ ] System.Security.Claims.ClaimsIdentity
- [ ] System.Security.Claims.ClaimsPrincipal
- [ ] System.Security.Principal.GenericIdentity
- [ ] System.Security.Principal.GenericPrincipal
- [ ] System.Security.Principal.TokenAccessLevels
- [ ] System.Security.Principal.TokenImpersonationLevel
- [ ] System.Security.Principal.WindowsIdentity
- [ ] System.Security.Principal.TokenType
- [ ] System.Security.Principal.TokenInformationClass
- [ ] System.Security.Principal.WindowsBuiltInRole
- [ ] System.Security.Principal.WindowsPrincipal
- [ ] System.Security.Principal.IdentityNotMappedException
- [ ] System.Security.Policy.AllMembershipCondition
- [ ] System.Security.Policy.ApplicationDirectory
- [ ] System.Security.Policy.ApplicationDirectoryMembershipCondition
- [ ] System.Security.Policy.ApplicationTrust
- [ ] System.Security.Policy.CodeGroup
- [ ] System.Security.Policy.Evidence
- [ ] System.Security.Policy.EvidenceBase
- [ ] System.Security.Policy.FileCodeGroup
- [ ] System.Security.Policy.FirstMatchCodeGroup
- [ ] System.Security.Policy.CodeConnectAccess
- [ ] System.Security.Policy.NetCodeGroup
- [ ] System.Security.Policy.PermissionRequestEvidence
- [ ] System.Security.Policy.PolicyException
- [ ] System.Security.Policy.PolicyLevel
- [ ] System.Security.Policy.PolicyStatementAttribute
- [ ] System.Security.Policy.PolicyStatement
- [ ] System.Security.Policy.Site
- [ ] System.Security.Policy.SiteMembershipCondition
- [ ] System.Security.Policy.StrongName
- [ ] System.Security.Policy.StrongNameMembershipCondition
- [ ] System.Security.Policy.UnionCodeGroup
- [ ] System.Security.Policy.Url
- [ ] System.Security.Policy.UrlMembershipCondition
- [ ] System.Security.Policy.Zone
- [ ] System.Security.Policy.ZoneMembershipCondition
- [ ] System.Security.Policy.GacInstalled
- [ ] System.Security.Policy.GacMembershipCondition
- [ ] System.Security.Policy.Hash
- [ ] System.Security.Policy.HashMembershipCondition
- [ ] System.Security.Policy.Publisher
- [ ] System.Security.Policy.PublisherMembershipCondition
- [ ] System.Resources.MissingManifestResourceException
- [ ] System.Resources.ResourceManager
- [ ] System.Diagnostics.StackTrace
- [ ] System.Diagnostics.StackFrame
- [ ] System.Diagnostics.Tracing.EventSourceException
- [ ] System.Threading.ThreadAbortException
- [ ] System.Reflection.MethodSemanticsAttributes
- [ ] System.Reflection.Emit.OpCodeType
- [ ] System.Reflection.Emit.StackBehaviour
- [ ] System.Reflection.Emit.OperandType
- [ ] System.Reflection.Emit.FlowControl
- [ ] System.Reflection.Emit.PackingSize
- [ ] System.Runtime.Serialization.SerializationException
- [ ] System.Runtime.Serialization.StreamingContext
- [ ] System.Runtime.Serialization.StreamingContextStates
- [ ] System.Text.BaseCodePageEncoding
- [ ] System.Text.DecoderNLS
- [ ] System.Text.InternalDecoderBestFitFallback
- [ ] System.Text.EncoderNLS
- [ ] System.Text.InternalEncoderBestFitFallback
- [ ] System.Text.EncodingNLS
- [ ] System.Text.ISCIIEncoding
- [ ] System.Text.DBCSCodePageEncoding
- [ ] System.Text.SBCSCodePageEncoding
- [ ] System.Text.GB18030Encoding
- [ ] System.Text.ISO2022Encoding
- [ ] System.Text.EUCJPEncoding
- [ ] System.ComponentModel.DataAnnotations.ValidationException
- [ ] System.IO.HandleInheritability
- [ ] System.Dynamic.DynamicObject
- [ ] System.Security.Cryptography.CngAlgorithm
- [ ] System.Security.Cryptography.CngAlgorithmGroup
- [ ] System.Security.Cryptography.CngKeyBlobFormat
- [ ] System.Security.Cryptography.CngProvider
- [ ] System.Data.ProviderBase.DbConnectionPoolIdentity
- [ ] System.Uri
- [ ] System.UriFormatException
- [ ] System.Net.HttpWebRequest
- [ ] System.Net.HttpWebResponse
- [ ] System.Net.ProtocolViolationException
- [ ] System.Net.WebException
- [ ] System.Net.WebHeaderCollection
- [ ] System.Net.WebResponse
- [ ] System.Net.WebSockets.WebSocketException
- [ ] System.Net.NetworkInformation.NetworkInformationException
- [ ] System.Net.Sockets.SocketException
- [ ] System.Security.Authentication.AuthenticationException
- [ ] System.Security.Authentication.InvalidCredentialException
- [ ] System.Security.Authentication.ExtendedProtection.ExtendedProtectionPolicy
- [ ] System.Security.Authentication.ExtendedProtection.ServiceNameCollection
- [ ] System.Security.Cryptography.X509Certificates.X509Certificate2
- [ ] System.Security.Cryptography.X509Certificates.X509CertificateCollection
- [ ] System.Security.Permissions.TypeDescriptorPermissionFlags
- [ ] System.Security.Permissions.TypeDescriptorPermission
- [ ] System.Threading.BarrierPostPhaseException
- [ ] System.Runtime.Versioning.FrameworkName
- [ ] System.IO.InvalidDataException
- [ ] System.ComponentModel.InvalidAsynchronousStateException
- [ ] System.CodeDom.Compiler.TempFileCollection
- [ ] System.Text.RegularExpressions.Capture
- [ ] System.Text.RegularExpressions.CaptureCollection
- [ ] System.Text.RegularExpressions.Group
- [ ] System.Text.RegularExpressions.GroupCollection
- [ ] System.Text.RegularExpressions.Match
- [ ] System.Text.RegularExpressions.MatchCollection
- [ ] System.Net.Http.HttpRequestException
- [ ] System.Numerics.BigInteger
- [ ] System.Numerics.Complex
- [ ] Windows.UI.Xaml.LayoutCycleException
- [ ] Windows.UI.Xaml.Markup.XamlParseException
- [ ] Windows.UI.Xaml.Automation.ElementNotAvailableException
- [ ] Windows.UI.Xaml.Automation.ElementNotEnabledException
- [ ] System.ServiceProcess.TimeoutException
- [ ] System.Xml.XmlException
- [ ] System.Xml.XmlQualifiedName
- [ ] System.Xml.XPath.XPathException
- [ ] System.Xml.Linq.XName

**Comment by stephentoub** [comment]
After these PRs:
https://github.com/dotnet/corefx/pull/12736
https://github.com/dotnet/corefx/pull/12762
https://github.com/dotnet/corefx/pull/12781
the following ~30 types across corefx still need [Serializable] applied (and in some cases the associated serialization implementations).  I skipped these as the projects haven't yet been updated for netstandard1.7 (cc: @ericstj):

Microsoft.CSharp
- [ ] Microsoft.CSharp.RuntimeBinder.RuntimeBinderException
- [ ] Microsoft.CSharp.RuntimeBinder.RuntimeBinderInternalCompilerException

System.ComponentModel.Annotations
- [ ] System.ComponentModel.DataAnnotations.ValidationException

System.Dynamic.Runtime
- [ ] System.Dynamic.DynamicObject

System.IO.FileSystem.DriveInfo
- [ ] System.IO.DriveInfo
- [ ] System.IO.DriveNotFoundException
- [ ] System.IO.DriveType

System.Net.Http
- [ ] System.Net.Http.HttpRequestException

System.Net.NetworkInformation
- [ ] System.Net.NetworkInformation.NetworkInformationException

System.Net.Primitives
- [ ] System.Net.Sockets.SocketException

System.Net.WebHeaderCollection
- [ ] System.Net.WebHeaderCollection

System.Net.WebSockets
- [ ] System.Net.WebSockets.WebSocketException

System.Runtime.Numerics
- [ ] System.Numerics.BigInteger
- [ ] System.Numerics.Complex

System.Security.AccessControl
- [ ] System.Security.AccessControl.PrivilegeNotHeldException

System.Security.Claims
- [ ] System.Security.Claims.Claim
- [ ] System.Security.Claims.ClaimsIdentity
- [ ] System.Security.Claims.ClaimsPrincipal

System.Security.Principal
- [ ] System.Security.Principal.GenericIdentity
- [ ] System.Security.Principal.GenericPrincipal
- [ ] System.Security.Principal.TokenImpersonationLevel

System.ServiceProcess.ServiceController
- [ ] System.ServiceProcess.TimeoutException

System.Text.Encoding.CodePages
- [ ] System.Text.BaseCodePageEncoding
- [ ] System.Text.DBCSCodePageEncoding
- [ ] System.Text.DecoderNLS
- [ ] System.Text.EncoderNLS
- [ ] System.Text.EncodingNLS
- [ ] System.Text.EUCJPEncoding
- [ ] System.Text.GB18030Encoding
- [ ] System.Text.InternalDecoderBestFitFallback
- [ ] System.Text.InternalEncoderBestFitFallback
- [ ] System.Text.ISCIIEncoding
- [ ] System.Text.ISO2022Encoding
- [ ] System.Text.SBCSCodePageEncoding

**Comment by danmosemsft** [comment]
@ericstj were you planning to do a bulk upgrade to 1.7? if not then we should probably open a separate issue for that.

**Comment by ericstj** [comment]
@danmosemsft I've been updating all the "hard" packages then individual ones as folks request.

I did write some scripts for bulk upgrade but the result wasn't pretty.  Too many special cases and places cleanup.  I'll help out here by upgrading all of these.  Feel free to send folks my way if they encounter additional libraries that still need updated.

> System.Net.Primitives

This is already updated to netstandard1.7.

**Comment by danmosemsft** [comment]
thanks @ericstj ..appreciate it.

**Comment by stephentoub** [comment]
Thank you, @ericstj.

> System.Net.Primitives

Ah, right, that one was upgraded, but I was blocked from making the change due to waiting for updated packages.

**Comment by ericstj** [comment]
@stephentoub did the commit I shared a couple days ago unblock you for HttpClient: https://github.com/ericstj/corefx/commit/38a31ba9e17f1fa6a86a8edd221771f1b8950f24 
Let me know if you'd like me to fix the tests and push this through.

**Comment by stephentoub** [comment]
> @stephentoub did the commit I shared a couple days ago unblock you for HttpClien

Sorry, I hadn't tried it as I thought you were still working on it (e.g. the tests).  If you could push it through, that'd be great.

**Comment by ericstj** [comment]
Sure, I'll pick that back up.

**Comment by ericstj** [comment]
All these are either done or in PR.

**Comment by stephentoub** [comment]
Thanks a bunch, Eric.
